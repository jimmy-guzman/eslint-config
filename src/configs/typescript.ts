import { config, configs } from "typescript-eslint";

import type { TypescriptOptions } from "../types";

import { GLOB_JS, GLOB_JSX, GLOB_TESTS } from "../constants";

export const typescriptConfig = (options: TypescriptOptions) => {
  return config(
    ...configs.strictTypeChecked,
    ...configs.stylisticTypeChecked.filter((config) => {
      return config.name === "typescript-eslint/stylistic-type-checked";
    }),
    {
      languageOptions: {
        parserOptions: {
          project: options.project,
          tsconfigRootDir: process.cwd(),
        },
      },
      name: "jimmy.codes/typescript",
      rules: {
        "@typescript-eslint/consistent-type-exports": [
          "error",
          { fixMixedExportsWithInlineTypeSpecifier: false },
        ],
        "@typescript-eslint/consistent-type-imports": [
          "error",
          { fixStyle: "separate-type-imports" },
        ],
        "@typescript-eslint/no-deprecated": "warn",
        "@typescript-eslint/no-misused-promises": [
          "error",
          { checksVoidReturn: { attributes: false } },
        ],
        "@typescript-eslint/restrict-template-expressions": [
          "error",
          { allowNumber: true },
        ],
      },
    },
    {
      files: [GLOB_JS, GLOB_JSX],
      ...configs.disableTypeChecked,
    },
    {
      files: GLOB_TESTS,
      name: "jimmy.codes/typescript/testing",
      rules: {
        "@typescript-eslint/no-unsafe-argument": "off",
        "@typescript-eslint/no-unsafe-assignment": "off",
      },
    },
  );
};
