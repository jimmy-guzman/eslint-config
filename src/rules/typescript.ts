import type { Rules } from "../types";

export const typescriptRules = {
  "@typescript-eslint/consistent-type-exports": [
    "error",
    { fixMixedExportsWithInlineTypeSpecifier: false },
  ],
  "@typescript-eslint/consistent-type-imports": [
    "error",
    { fixStyle: "separate-type-imports" },
  ],
  "@typescript-eslint/no-deprecated": "warn",
  "@typescript-eslint/no-misused-promises": [
    "error",
    { checksVoidReturn: { attributes: false } },
  ],
  "@typescript-eslint/no-unnecessary-type-conversion": "error",
  "@typescript-eslint/no-unused-vars": [
    "error",
    // https://typescript-eslint.io/rules/no-unused-vars/#benefits-over-typescript
    {
      args: "all",
      argsIgnorePattern: "^_",
      caughtErrors: "all",
      caughtErrorsIgnorePattern: "^_",
      destructuredArrayIgnorePattern: "^_",
      ignoreRestSiblings: true,
      varsIgnorePattern: "^_",
    },
  ],
  "@typescript-eslint/no-use-before-define": [
    "error",
    {
      allowNamedExports: false,
      classes: false,
      functions: false,
      variables: true,
    },
  ],
  "@typescript-eslint/restrict-template-expressions": [
    "error",
    { allowNumber: true },
  ],
  "no-use-before-define": "off",
} satisfies Rules;
