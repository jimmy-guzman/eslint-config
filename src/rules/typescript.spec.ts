import { typescriptRules } from "./typescript";

test("should create typescript rules", () => {
  expect(typescriptRules).toMatchSnapshot();
});
