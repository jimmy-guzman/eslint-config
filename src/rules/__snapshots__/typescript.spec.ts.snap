// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create typescript rules 1`] = `
{
  "@typescript-eslint/consistent-type-exports": [
    "error",
    {
      "fixMixedExportsWithInlineTypeSpecifier": false,
    },
  ],
  "@typescript-eslint/consistent-type-imports": [
    "error",
    {
      "fixStyle": "separate-type-imports",
    },
  ],
  "@typescript-eslint/no-deprecated": "warn",
  "@typescript-eslint/no-misused-promises": [
    "error",
    {
      "checksVoidReturn": {
        "attributes": false,
      },
    },
  ],
  "@typescript-eslint/no-unnecessary-type-conversion": "error",
  "@typescript-eslint/no-unused-vars": [
    "error",
    {
      "args": "all",
      "argsIgnorePattern": "^_",
      "caughtErrors": "all",
      "caughtErrorsIgnorePattern": "^_",
      "destructuredArrayIgnorePattern": "^_",
      "ignoreRestSiblings": true,
      "varsIgnorePattern": "^_",
    },
  ],
  "@typescript-eslint/no-use-before-define": [
    "error",
    {
      "allowNamedExports": false,
      "classes": false,
      "functions": false,
      "variables": true,
    },
  ],
  "@typescript-eslint/restrict-template-expressions": [
    "error",
    {
      "allowNumber": true,
    },
  ],
  "no-use-before-define": "off",
}
`;
